File: qa/extract-workbook.ts
import { mkdir, writeFile } from "node:fs/promises";
import { join } from "node:path";
import { inflateRawSync } from "node:zlib";

interface SheetManifest {
  name: string;
  rows: Array<Record<string, string>>;
}

function columnIndex(label: string) {
  return [...label].reduce((value, char) => value * 26 + (char.charCodeAt(0) - 64), 0) - 1;
}

async function readArchiveEntries(path: string) {
  const { readFile } = await import("node:fs/promises");
  const buffer = await readFile(path);
  const text = buffer.toString("binary");
  const entries = new Map<string, Buffer>();
  let offset = 0;

  while (offset < buffer.length) {
    const signature = buffer.readUInt32LE(offset);

    if (signature !== 0x04034b50) {
      break;
    }

    const compressionMethod = buffer.readUInt16LE(offset + 8);
    const compressedSize = buffer.readUInt32LE(offset + 18);
    const fileNameLength = buffer.readUInt16LE(offset + 26);
    const extraLength = buffer.readUInt16LE(offset + 28);
    const fileName = text.slice(offset + 30, offset + 30 + fileNameLength);
    const dataStart = offset + 30 + fileNameLength + extraLength;
    const dataEnd = dataStart + compressedSize;
    const entry = buffer.subarray(dataStart, dataEnd);
    entries.set(fileName, compressionMethod === 8 ? inflateRawSync(entry) : entry);
    offset = dataEnd;
  }

  return entries;
}

function extractSharedStrings(xml: string) {
  return [...xml.matchAll(/<t[^>]*>([\s\S]*?)<\/t>/g)].map((match) =>
    match[1]
      .replace(/&amp;/g, "&")
      .replace(/&lt;/g, "<")
      .replace(/&gt;/g, ">")
  );
}

function extractRows(xml: string, sharedStrings: string[]) {
  const rows = [...xml.matchAll(/<row[\s\S]*?<\/row>/g)].map((match) => match[0]);
  return rows.map((row) => {
    const cells = [...row.matchAll(/<c[^>]*r="([A-Z]+)\d+"[^>]*?(?:t="([^"]+)")?[^>]*>([\s\S]*?)<\/c>/g)];
    const values = new Map<number, string>();

    cells.forEach((cell) => {
      const [, column, type, body] = cell;
      const shared = body.match(/<v>(\d+)<\/v>/);
      const inline = body.match(/<t[^>]*>([\s\S]*?)<\/t>/);
      const value =
        type === "s" && shared
          ? sharedStrings[Number(shared[1])] ?? ""
          : inline?.[1]?.replace(/&amp;/g, "&").replace(/&lt;/g, "<").replace(/&gt;/g, ">") ?? body.match(/<v>([\s\S]*?)<\/v>/)?.[1] ?? "";
      values.set(columnIndex(column), value);
    });

    const headers = Array.from({ length: Math.max(...values.keys(), 0) + 1 }, (_, index) => values.get(index) ?? "");
    return headers;
  });
}

async function main() {
  const root = "/Users/tanmaykumar/Downloads/RuleMind.AI";
  const workbookPath = join(root, "Test Cases", "rulemind_test_suite.xlsx");
  const resultsDir = join(root, "qa", "results");
  await mkdir(resultsDir, { recursive: true });

  const entries = await readArchiveEntries(workbookPath);
  const workbookXml = entries.get("xl/workbook.xml")?.toString("utf8") ?? "";
  const relsXml = entries.get("xl/_rels/workbook.xml.rels")?.toString("utf8") ?? "";
  const sharedStrings = extractSharedStrings(entries.get("xl/sharedStrings.xml")?.toString("utf8") ?? "");

  const relationshipMap = new Map(
    [...relsXml.matchAll(/Relationship[^>]*Id="([^"]+)"[^>]*Target="([^"]+)"/g)].map((match) => [match[1], match[2].replace(/^\//, "")])
  );

  const sheets = [...workbookXml.matchAll(/sheet[^>]*name="([^"]+)"[^>]*r:id="([^"]+)"/g)].map((match) => ({
    name: match[1],
    target: relationshipMap.get(match[2]) ?? ""
  }));

  const manifest: SheetManifest[] = sheets.map((sheet) => {
    const rows = extractRows(entries.get(sheet.target)?.toString("utf8") ?? "", sharedStrings);
    const headers = rows[0] ?? [];
    return {
      name: sheet.name,
      rows: rows.slice(1).map((row) =>
        Object.fromEntries(headers.map((header, index) => [header || `col_${index + 1}`, row[index] ?? ""]))
      )
    };
  });

  await writeFile(
    join(resultsDir, "workbook-manifest.json"),
    JSON.stringify(
      {
        generatedAt: new Date().toISOString(),
        sheets: manifest
      },
      null,
      2
    ),
    "utf8"
  );

  process.stdout.write(`Workbook manifest written to ${join(resultsDir, "workbook-manifest.json")}\n`);
}

main().catch((error) => {
  process.stderr.write(`${String(error)}\n`);
  process.exitCode = 1;
});

File: qa/extract-workbook.py
from __future__ import annotations

import json
from pathlib import Path
from xml.etree import ElementTree as ET
from zipfile import ZipFile


NS = {
    "a": "http://schemas.openxmlformats.org/spreadsheetml/2006/main",
    "r": "http://schemas.openxmlformats.org/officeDocument/2006/relationships",
}


def column_to_index(label: str) -> int:
    value = 0
    for char in label:
        value = value * 26 + (ord(char) - 64)
    return value - 1


def extract_rows(sheet_xml: bytes, shared_strings: list[str]) -> list[list[str]]:
    worksheet = ET.fromstring(sheet_xml)
    rows: list[list[str]] = []

    for row in worksheet.findall(".//a:sheetData/a:row", NS):
        values: dict[int, str] = {}
        for cell in row.findall("a:c", NS):
            ref = cell.attrib["r"]
            col = "".join(char for char in ref if char.isalpha())
            cell_type = cell.attrib.get("t")
            value_node = cell.find("a:v", NS)
            value = ""

            if cell_type == "s" and value_node is not None:
                value = shared_strings[int(value_node.text or "0")]
            elif cell_type == "inlineStr":
                inline = cell.find("a:is", NS)
                if inline is not None:
                    value = "".join(text.text or "" for text in inline.iter("{http://schemas.openxmlformats.org/spreadsheetml/2006/main}t"))
            elif value_node is not None:
                value = value_node.text or ""

            values[column_to_index(col)] = value

        if values:
            max_col = max(values)
            rows.append([values.get(index, "") for index in range(max_col + 1)])

    return rows


def main() -> None:
    root = Path("/Users/tanmaykumar/Downloads/RuleMind.AI")
    workbook_path = root / "Test Cases" / "rulemind_test_suite.xlsx"
    results_dir = root / "qa" / "results"
    results_dir.mkdir(parents=True, exist_ok=True)

    with ZipFile(workbook_path) as archive:
        workbook = ET.fromstring(archive.read("xl/workbook.xml"))
        relationships = ET.fromstring(archive.read("xl/_rels/workbook.xml.rels"))
        relationship_map = {rel.attrib["Id"]: rel.attrib["Target"].lstrip("/") for rel in relationships}
        shared_strings: list[str] = []

        if "xl/sharedStrings.xml" in archive.namelist():
            sst = ET.fromstring(archive.read("xl/sharedStrings.xml"))
            for item in sst:
                shared_strings.append(
                    "".join(text.text or "" for text in item.iter("{http://schemas.openxmlformats.org/spreadsheetml/2006/main}t"))
                )

        manifest = []
        for sheet in workbook.find("a:sheets", NS):
            target = relationship_map[sheet.attrib["{http://schemas.openxmlformats.org/officeDocument/2006/relationships}id"]]
            rows = extract_rows(archive.read(target), shared_strings)
            headers = rows[0] if rows else []
            manifest.append(
                {
                    "name": sheet.attrib["name"],
                    "rows": [
                        {headers[index] if headers[index] else f"col_{index + 1}": value for index, value in enumerate(row)}
                        for row in rows[1:]
                    ],
                }
            )

    output_path = results_dir / "workbook-manifest.json"
    output_path.write_text(
        json.dumps(
            {
                "generatedAt": __import__("datetime").datetime.utcnow().isoformat(timespec="seconds") + "Z",
                "sheets": manifest,
            },
            indent=2,
        ),
        encoding="utf-8",
    )
    print(f"Workbook manifest written to {output_path}")


if __name__ == "__main__":
    main()
